### ISSUE TITLE
Bug: Doesn't populate existing objects

### ISSUE THREAD

**Issue by jilleJr** [jilleJr/Newtonsoft.Json-for-Unity.Converters#20]
## Description

At time of writing the PartialConverters always creates new objects instead of trying to populate existing ones. This should be fixed.

Given the snippet below:

```cs
object boxedVector3 = new Vector3(1, 2, 3);
string json = @"{ ""x"": 4 }";
JsonConvert.PopulateObject(json, boxedVector3);

Console.WriteLine(boxedVector3);
```

### Expected behavior

Output: `(4, 2, 3)`

### Actual behavior

Output: `(4, 0, 0)`

## Details

Host machine OS running Unity Editor 👉 Windows

Unity build target 👉 Editor

Newtonsoft.Json-for-Unity package version 👉 12.0.301

Newtonsoft.Json-for-Unity.Converters package version 👉 x.1.0-preview

I was using Unity version 👉 2019.2.11f1

## Additional info

The PartialConverter does receive the old value `(1, 2, 3)`, so it should be able to use that as initial value for the array. Shouldn't be too difficult.

**Comment by jilleJr** [comment]
Wait, false alarm.

This does work as expected. The value does get set.

**Comment by jilleJr** [comment]
Wait, true alarm.

Found a veery similar issue and reused this issue instead.

**Comment by nialna** [comment]
I'd like to try fixing this one if possible, is there anything important I should know before I try?

**Comment by jilleJr** [comment]
@nialna Oh that's very kind of you!
Most important thing to keep in mind is to write some tests to prove that it works. I'm not that nitpicky about how you write your tests. As long as there's something.

Also I've so far developed the repo with Unity 2018.4.14f1. Try to use the same or similar version because Unity is backwards compatible with it's assets but not forward compatible.

Other than that you can find the package source in the Packages folder and the tests source in the Assets folder.

Just give it a try and create a pull request when you feel done; I'll try to be merciful in reviewing it ;) Good luck ^^

**Comment by jilleJr** [comment]
I'm assigning you to get a better overview so no one (me in particular) starts working on it at the same time. No need to stress over it, as far as I know you're the only one waiting for this feature.

**Comment by nialna** [comment]
I've been looking at this a bit and I'm slightly confused by how it works and could do with a bit of help to figure out where to actually make changes.

From what I've found so far, it seems the key function is `InternalReadJson` in `PartialConverter`, which will eventually call `CreateInstanceFromValues`. I guess for most types, what you'd want to do is instead use the existing value if it exists, and apply the values in the `ValuesArray`?

But then what I don't understand is that `CreateInstanceFromValues` is only implemented for the converters that exist in the repo, and for example doesn't exist for `ScriptableObject`. What I've found is that those ones are created in `createObjectContract`, but then I really have no idea how that works with the partial converters?

I'm probably misunderstanding a lot of stuff because it's a bit hard to figure out how everything works from the code, so sorry if I'm not making sense

**Comment by jilleJr** [comment]
@nialna Ah yes. Let me get back to you once I gathered some time to investigate. Cannot promise today though, but probably in the weekend.

**Comment by jilleJr** [comment]
@nialna Excuse my delay, but yea so for ScriptableObject I'm just relying on the default implementation of Newtonsoft.Json to handle the population on existing type. I've not added any custom converter for ScriptableObjects as it just resulted in weird hacks when I tried because it needs to handle user-defined fields, so for ScriptableObjects in particular I just added a custom ContractResolver.

For all types that has a custom converter, if I remember correctly, all should derive from `PartialConverter<T, TInner>`. And currently it does not use the `existingObject` argument at all. (As can be seen here:)

https://github.com/jilleJr/Newtonsoft.Json-for-Unity.Converters/blob/d008ac078cbd2fb81e166028a5edee0ff796f748/Packages/Newtonsoft.Json-for-Unity.Converters/UnityConverters/PartialConverter.cs#L132-L142

So what I had in mind when I initially wrote this issue was to change it the following:

1. Write test in `TypeTester<T>` for asserting that the current converters implementation doesn't populate existing ones correctly. The "expected/actual" test above in the issue description should be a good starting ground.
2. Propegate the `existingObject` down to the `InternalReadJson`
3. Add some "create new instance" method that the converters has to implement, then use that if the `existingObject` is null.
4. In the `ReadValue` pass in the `existingObject` instead, and change the method to instead apply the value to the object directly instead. (probably need to pass the `existingObject` by reference for structs to update correctly). This all instead of populating a temporary ValuesArray.
5. Return the `existingObject` instead, making `CreateInstanceFromValues` obsolete (not necessarily with the Obsolete attribute, just that it's possible to remove it).

All in all, replacing the use of the temporary `ValuesArray` with an actual instance of the desired type (quite more performant as well), and in the result also fixing this PopulateObject bug.

That's how I would do it. You're well allowed to remix my suggestion however you want.
